pkg/sdk/functions_def.go

```diff
@@ -0,0 +1,410 @@
+	OptionalTextAssignment("COMMENT", g.ParameterOptions().SingleQuotes()).
+	OptionalSQL("SECURE")
+
+var functionUnset = g.NewQueryStruct("FunctionUnset").
+	OptionalSQL("SECURE").
+	OptionalSQL("COMMENT").
+	OptionalSQL("LOG_LEVEL").
+	OptionalSQL("TRACE_LEVEL")
+
+var (
+	functionNullOrNot     = g.NewQueryStruct("FunctionNullOrNot").OptionalSQL("NULL").OptionalSQL("NOT NULL")
+	functionStrictOrNot   = g.NewQueryStruct("FunctionStrictOrNot").OptionalSQL("STRICT").OptionalSQL("CALLED ON NULL INPUT")
+	functionVolatileOrNot = g.NewQueryStruct("FunctionVolatileOrNot").OptionalSQL("VOLATILE").OptionalSQL("IMMUTABLE")
+	functionImports       = g.NewQueryStruct("FunctionImports").Text("Import", g.KeywordOptions().SingleQuotes())
+	functionPackages      = g.NewQueryStruct("FunctionPackages").Text("Package", g.KeywordOptions().SingleQuotes())
+	functionDefinition    = g.NewQueryStruct("FunctionDefinition").Text("Definition", g.KeywordOptions())
+)
+
+var FunctionsDef = g.NewInterface(
+	"Functions",
+	"Function",
+	g.KindOfT[SchemaObjectIdentifier](),
+).CustomOperation(
+	"CreateFunctionForJava",
+	"https://docs.snowflake.com/en/sql-reference/sql/create-function",
+	g.NewQueryStruct("CreateFunctionForJava").
+		Create().
+		OrReplace().
+		OptionalSQL("TEMPORARY").
+		OptionalSQL("SECURE").
+		SQL("FUNCTION").
+		IfNotExists().
+		Name().
+		ListQueryStructField(
+			"Arguments",
+			functionArgument,
+			g.ParameterOptions().Parentheses().NoEquals()).
+		OptionalSQL("COPY GRANTS").
+		OptionalQueryStructField(
+			"Returns",
+			functionReturns,
+			g.KeywordOptions().SQL("RETURNS"),
+		).
+		OptionalQueryStructField(
+			"NullOrNot",
+			functionNullOrNot,
+			g.KeywordOptions(),
+		).
+		SQL("LANGUAGE JAVA").
+		OptionalQueryStructField(
+			"StrictOrNot",
+			functionStrictOrNot,
+			g.KeywordOptions(),
+		).
+		OptionalQueryStructField(
+			"VolatileOrNot",
+			functionVolatileOrNot,
+			g.KeywordOptions(),
+		).
+		OptionalTextAssignment("RUNTIME_VERSION", g.ParameterOptions().SingleQuotes()).
+		OptionalTextAssignment("COMMENT", g.ParameterOptions().SingleQuotes()).
+		ListQueryStructField(
+			"Imports",
+			functionImports,
+			g.ParameterOptions().Parentheses().SQL("IMPORTS"),
+		).
+		ListQueryStructField(
+			"Packages",
+			functionPackages,
+			g.ParameterOptions().Parentheses().SQL("PACKAGES"),
+		).
+		OptionalTextAssignment("HANDLER", g.ParameterOptions().SingleQuotes()).
+		ListAssignment("EXTERNAL_ACCESS_INTEGRATIONS", "AccountObjectIdentifier", g.ParameterOptions().Parentheses()).
+		ListQueryStructField(
+			"Secrets",
+			functionSecret,
+			g.ParameterOptions().Parentheses().SQL("SECRETS"),
+		).
+		OptionalTextAssignment("TARGET_PATH", g.ParameterOptions().SingleQuotes()).
+		OptionalQueryStructField(
+			"FunctionDefinition",
+			functionDefinition,
+			g.ParameterOptions().NoEquals().SingleQuotes().SQL("AS"),
+		).
+		WithValidation(g.ValidIdentifier, "name"),
```

Comment:
no validation that "imports" are illegal with "packages". Some in other operations

Reply:
not sure if imports and packages really are incompatible with each other. It just says: 

> For a dependency you specify with PACKAGES, you do not need to also specify its JAR file in an IMPORTS clause.

Perhaps we are going overboard on the validation requirements. Snowflake also does its own internal validation of SQL commands, so if this truly was a problem, it would be thrown back to the user anyways.

Reply:
maybe I missed something in the first one, but I have a feeling that sometimes Snowflake errors confuse people and they think that there's something wrong with the Terraform provider, but ok, maybe we can leave it as is because more sophisticated validation should be done on the Schema level for those kinds of validations